cmd/snapshotgc/snapshotgc.go

```diff
@@ -0,0 +1,61 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	applicationapiv1alpha1 "github.com/redhat-appstudio/application-api/api/v1alpha1"
+	releasev1alpha1 "github.com/redhat-appstudio/release-service/api/v1alpha1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var (
+	scheme = runtime.NewScheme()
+)
+
+func init() {
+	applicationapiv1alpha1.AddToScheme(scheme)
+	clientgoscheme.AddToScheme(scheme)
+	releasev1alpha1.AddToScheme(scheme)
+}
+
+type snapshotData struct {
+	snapshot applicationapiv1alpha1.Snapshot
+	// environmentBinding applicationapiv1alpha1.SnapshotEnvironmentBinding
+	release releasev1alpha1.Release
+}
+
+func getSnapshotsForReleases(
+	cl client.Client,
+	snapToData map[string]snapshotData,
+) map[string]snapshotData {
+	releases := &releasev1alpha1.ReleaseList{}
+	err := cl.List(context.Background(), releases)
```

Comment:
I don't like that we are fetching all snapshots at once. It generates a big load on api and also we have to store a lot of results in memory.

Why we don't go namespace by namespace, one by one, process one namespace, free the memory and continue with another. It will also simplify code it won't need to keep complex key in map

Reply:
I'll look into that. I was considering that initially, but had something in mind that made me prefer looking at it as a flat list.

One thing in favor of the current approach is the number of API calls.
Suppose that we go namespace by namespace, it means that we'll have to fetch releases (and bindings later) for each namespace (also namespaces that don't have them). 
Do you expect memory to be more of an issue than the total number of calls?

Reply:
True is somewhere in the middle, you don;t want many calls, but you also don't want huge calls, where you trigger OOM attack on server by listing the whole world.

But you are afraid of too many api calls, why do you then fetch snapshots one by one instead of getting all snapshots and then process it.

The middle road for me would be:
Per namespace: 
- fetch all snapshots at once
- fetch all releases at once
- fetch all SEBs (bindings) at once
Do magic and find which snapshots should be deleted, and delete them


Integration service in prod is currently using 650MB of memory, where the majority is caused by caches of CRs. 600MB sounds quite a lot for me for pruner.

Reply:
ok. I changed the implementation.

- Going namespace by namespace
- Storing only the snapshot name at the first step
- Will fetch the snapshot itself only later in the process, and only for those which are not associated with release/binding.

Reply:
@MartinBasti let me know if you're ok with the change. I'll apply the same approach for coming PRs.

Reply:
Looks good to me, thank you